File: src/main/java/de/tectoast/emolga/utils/GPIOManager.java
package de.tectoast.emolga.utils;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

public class GPIOManager {

    private static final Logger logger = LoggerFactory.getLogger(GPIOManager.class);

    private static final ScheduledExecutorService service = Executors.newScheduledThreadPool(1);

    private static final int TURN_ON_TIME = 500;
    private static final int TURN_OFF_TIME = 500;
    private static final int POWER_OFF = 5000;

    private static void toggle(PC pc, int duration) {
        exec(new String[]{"/usr/bin/gpio", "write", pc.getWritePin(), "0"});
        service.schedule(() -> exec(new String[]{"/usr/bin/gpio", "write", pc.getWritePin(), "1"}), duration, TimeUnit.MILLISECONDS);
    }

    public static void startServer(PC pc) {
        toggle(pc, TURN_ON_TIME);
    }

    public static void stopServer(PC pc) {
        toggle(pc, TURN_OFF_TIME);
    }

    public static void powerOff(PC pc) {
        toggle(pc, POWER_OFF);
    }

    private static void exec(String[] cmd) {
        try {
            Runtime.getRuntime().exec(cmd);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static boolean isOn(PC pc) throws IOException {
        return new BufferedReader(new InputStreamReader(Runtime.getRuntime().exec(new String[]{"/usr/bin/gpio", "read", pc.getReadPin()}).getInputStream())).readLine().trim().equals("1");
    }

    public enum PC {
        FLORIX_2(2, 24, 964571226964115496L),
        FLORIX_3(3, 25, 975076826588282962L);

        private final int writePin;
        private final int readPin;
        private final long messageId;

        PC(int writePin, int readPin, long messageId) {
            this.writePin = writePin;
            this.readPin = readPin;
            this.messageId = messageId;
        }

        public static PC byMessage(long messageId) {
            return Arrays.stream(values()).filter(pc -> pc.messageId == messageId).findFirst().orElse(null);
        }

        public String getWritePin() {
            return String.valueOf(writePin);
        }

        public String getReadPin() {
            return String.valueOf(readPin);
        }
    }

}

File: src/main/kotlin/de/tectoast/emolga/utils/GPIOManager.kt
@file:Suppress("unused")

package de.tectoast.emolga.utils

import io.ktor.client.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import java.io.BufferedReader
import java.io.InputStreamReader

private const val TURN_ON_TIME = 500
private const val TURN_OFF_TIME = 500
private const val POWER_OFF = 5000

abstract class GPIOManager(val boardType: BoardType) {
    abstract suspend fun startServer(pc: PC)
    abstract suspend fun stopServer(pc: PC)
    abstract suspend fun powerOff(pc: PC)
    abstract suspend fun isOn(pc: PC): Boolean


    companion object {
        private var instance: GPIOManager = RemoteGPIOManager(httpClient, "http://rpicontrol:8080")
        operator fun invoke(): GPIOManager = instance

        fun setGlobalManager(manager: GPIOManager) {
            instance = manager
        }
    }

}

class RemoteGPIOManager
/**
 *
 */
    (val client: HttpClient, val url: String) : GPIOManager(BoardType.BCM) {
    override suspend fun startServer(pc: PC) = push(pc, TURN_ON_TIME)

    override suspend fun stopServer(pc: PC) = push(pc, TURN_OFF_TIME)

    override suspend fun powerOff(pc: PC) = push(pc, POWER_OFF)

    private suspend fun push(pc: PC, delay: Int) {
        withContext(Dispatchers.IO) {
            client.post("$url/push/${pc.getWritePin(boardType)}") {
                setBody("$delay")
            }
        }
    }

    override suspend fun isOn(pc: PC) = withContext(Dispatchers.IO) {
        client.get("$url/status/${pc.getReadPin(boardType)}").bodyAsText().contains("level=0")
    }
}

object DirectGPIOManager : GPIOManager(BoardType.WPi) {
    private suspend fun toggle(pc: PC, duration: Int) {
        val writePin = pc.getWritePin(boardType)
        exec(arrayOf("/usr/bin/gpio", "write", writePin, "0"))
        delay(duration.toLong())
        exec(arrayOf("/usr/bin/gpio", "write", writePin, "1"))
    }

    override suspend fun startServer(pc: PC) {
        toggle(pc, TURN_ON_TIME)
    }

    override suspend fun stopServer(pc: PC) {
        toggle(pc, TURN_OFF_TIME)
    }

    override suspend fun powerOff(pc: PC) {
        toggle(pc, POWER_OFF)
    }

    private fun exec(cmd: Array<String>) {
        Runtime.getRuntime().exec(cmd)
    }

    override suspend fun isOn(pc: PC): Boolean {
        return withContext(Dispatchers.IO) {
            BufferedReader(
                InputStreamReader(
                    Runtime.getRuntime()
                        .exec(arrayOf("/usr/bin/gpio", "read", pc.getReadPin(boardType))).inputStream
                )
            ).readLine()
        }.trim() != "1"
    }

}

enum class PC(
    private val writePinWPi: Int,
    private val readPinWPi: Int,
    private val writePinBCM: Int,
    private val readPinBCM: Int
) {
    FLORIX_2(
        2,
        24,
        3,
        15,
    ),
    FLORIX_4(
        3,
        25,
        2,
        14,
    );

    fun getWritePin(type: BoardType): String {
        return when (type) {
            BoardType.WPi -> writePinWPi
            BoardType.BCM -> writePinBCM
        }.toString()
    }

    fun getReadPin(type: BoardType): String {
        return when (type) {
            BoardType.WPi -> readPinWPi
            BoardType.BCM -> readPinBCM
        }.toString()
    }
}

enum class BoardType {
    WPi, BCM
}
